### ISSUE TITLE
Can't figure out how to correctly read the mesh information of a PolyMesh and PolyMeshDetail - recast4j

### ISSUE THREAD

**Issue by TrentWDB** [recastnavigation/recastnavigation#238]
This may not be the correct place to post my issue, but I'm assuming that the data structures between recast4j and the cpp version are similar.

This is more of a request for help rather than an issue, but I've been trying to read and output the non-triangulated mesh data to an obj using recast4j.

What I have so far is this:

            for (int i = 0; i < polyMesh.nverts; i++) {
                // not sure this is right
                int xIndex = polyMesh.verts[i * 3];
                int yIndex = polyMesh.verts[i * 3 + 1];
                int zIndex = polyMesh.verts[i * 3 + 2];

                float x = polyMeshDetail.verts[xIndex];
                float y = polyMeshDetail.verts[yIndex];
                float z = polyMeshDetail.verts[zIndex];

                fw.write("v " + x + " " + y + " " + z + "\n");
            }

            for (int i = 0; i < polyMesh.npolys; i++) {
                int polygonIndex = i * 2 * polyMesh.nvp; // following their specs?
                int[] polygon = new int[polyMesh.nvp];

                // fill the polygon with the corresponding part in the polys array
                for (int indexOffset = 0; indexOffset < polyMesh.nvp; indexOffset++) {
                    int polygonVertexIndex = polygonIndex + indexOffset;
                    polygon[indexOffset] = polyMesh.polys[polygonVertexIndex];
                }

                fw.write("f ");
                for (int a = 0; a < polygon.length; a++) {
                    int vertexIndex = polygon[a];
                    if (vertexIndex == 65535) {
                        continue;
                    }

                    fw.write("" + (polygon[a] + 1));
                    if (a < polygon.length - 1) {
                        fw.write(" ");
                    }
                }
                fw.write("\n");
            }

Any help on this would be appreciated!

Attached is a picture of what the rendered obj looks like. (It at least renders so my code is half correct.)
![screenshot 2016-11-04 04 49 54](https://cloud.githubusercontent.com/assets/5142257/20001443/2bfb2f52-a24a-11e6-9647-ad1d826dc342.png)

**Comment by hymerman** [comment]
Looks like there's a bit of a difference in data structures! In the cpp version, each tile stores an array of detail poly descriptors, an array of detail poly verts (shared between polys), and an array of detail poly tris, and those descriptors index into parts of the verts and tris arrays.

Each tile stores this (amongst lots of other stuff):

    dtPolyDetail* detailMeshes;			///< The tile's detail sub-meshes. [Size: dtMeshHeader::detailMeshCount]
	
    /// The detail mesh's unique vertices. [(x, y, z) * dtMeshHeader::detailVertCount]
    float* detailVerts;	

    /// The detail mesh's triangles. [(vertA, vertB, vertC) * dtMeshHeader::detailTriCount]
    unsigned char* detailTris;	

Each detail poly looks like this:

    struct dtPolyDetail
    {
    	unsigned int vertBase;			///< The offset of the vertices in the dtMeshTile::detailVerts array.
    	unsigned int triBase;			///< The offset of the triangles in the dtMeshTile::detailTris array.
    	unsigned char vertCount;		///< The number of vertices in the sub-mesh.
    	unsigned char triCount;			///< The number of triangles in the sub-mesh.
    };

... So clearly iterating all detail tris is very very different from your code example. Sorry!

Despite being stored differently, it's likely that in both versions, the *detail* verts and tris are completely separate to the non-detail (we don't have a better name for that...) verts and tris. Your code looks odd because it's using normal indices as indices into the detail verts, which I think is going to produce odd results.

In other words, if you're after the detail mesh, you don't want to touch the poly mesh at all.

Hope that helps a little!

**Comment by TrentWDB** [comment]
Yeah that helped, thanks hymerman! As it turns out what was confusing me was that the verts array in the polymesh was an int array. I assumed that since it was an int array it must be indices pointing to the float coordinates inside the polymeshdetails verts array since it was the only float array in either data structure.

However, the int verts array in polymesh were actually the coordinates, and the scale to convert them to the proper float coordinates is the cell size and cell height. Hopefully this will help anyone in the future who runs into the same confusion that I did.

**Comment by hymerman** [comment]
You're very welcome! I've seen a similar thing in A* Pathfinding Project Pro, which implements parts of Recast in C#; integer coordinates scaled by a cell size.
